### PR TITLE
Path issue windows mlflow plugin

### PR DESCRIPTION
**Pull request by vthorey**
Changes
---------
-> Use PurePosixPath to manipulate paths in artifact folder on MLflow artifact_repository plugin

-> Make DSSManagedFolder.upload_folder work on windows
For this we used the string method replace to replace backslashes with normal slashes because
1. we work with relative path (so no prefix "C:/" for instance)
2. we don't know if pathlib is installed since we support python2.7 (whereas in the MLflow plugin part we need MLflow, hence python >= 3.6 where pathlib is builtin)

Testing
-------
Companion test PR: https://github.com/dataiku/dip/pull/17453 with the DOD there

To test on windows you can install the kit and directly edit the dataikuapi in the kit as suggested by @lpenet here: https://app.shortcut.com/dataiku/story/92017/experiment-tracking-does-not-work-on-windows

### CODE REVIEW

File: dataikuapi/dss/managedfolder.py [hunk 1]
```diff
@@ -123,7 +123,8 @@ def upload_folder(self, path, folder):
             for file in files:
                 filename = os.path.join(root, file)
                 with open(filename, "rb") as f:
-                    self.put_file(os.path.join(path, os.path.relpath(filename, folder)), f)
+                    rel_posix_path = os.path.relpath(filename, folder).replace("\\", "/")
+                    self.put_file("{}/{}".format(path, rel_posix_path), f)
```
Comment (lpenet) [inline dataikuapi/dss/managedfolder.py:127]: This is not good on posix, as filenames _may_ contain those strange things that Humans who do not have the privilege to be geeks use, such as spaces and... \

You can do the following simple test:
- create  a `test_upload` dir containing a file named `a file_with\` (on GNU/Linux and OSX, you can do `touch a\ file_with\\` to create this one
- upload the dir using `upload_folder`
- see that the name of the uploaded file is `a file_with` :scream: 

Maybe we can perform the replacement iff `os.sep == '\'` ?

And, well, we should add a test for this case, in this PR (preferred) or in an additional test card, if there is not already one.

**Review by lpenet — CHANGES_REQUESTED** [review]
We have to be extra-cautious manipulating paths in `upload_folder`.
Still have to test the PR on Windows.

**Comment by vthorey** [comment]
@jenkins-dataiku 
```{
    "agents": {
        "agent": {
            "base-definition": "std1"
        }
    },
    "scenarios": {
        "inttests": {
            "scenario-dir": "ml/prediction/mlflow/experiment_tracking/",
            "enrichment": true,
            "preinstall": "install-python37;mlflow",
            "timeout": 300
        }
    }
}```

File: dataikuapi/dss/managedfolder.py (hunk 1 above)
Comment (vthorey) [reply dataikuapi/dss/managedfolder.py]: You are right, thank you for pointing this !

**Comment by lpenet** [comment]
To test experiment tracking is ok on windows:
- install DSS on windows following https://analytics.dataiku.com/projects/RDWIKI/wiki/1480/Setting%20up%20a%20Windows%20VM%20on%20Azure then installing  https://downloads.dataiku.com/public/launcher/windows/Dataiku%20Launcher%20Setup%201.3.5.exe
- replaced its dataikuapi dir with the content of the dataiku-api-client-python
- created a python 3.7 code env with core & jupyter packaged +
  - mlflow==1.21.0
  - scikit-learn>=0.20,<0.21
  - statsmodels>=0.10,<0.11
  - scipy>=1.2,<1.3
  - protobuf==3.16.0
- import the iris dataset and infer its schema (to have all cols as double excepted the class)
- prepare a "iris_prepared" dataset, removing rows where class is empty from "iris"
- create a project, then a notebook using the just created code env
- run this python code:
```
import dataiku
from sklearn import svm
client = dataiku.api_client()
project = client.get_default_project()
mlflow_extension = project.get_mlflow_extension()
mf = project.get_managed_folder("8rsHuleI")
with project.setup_mlflow(mf) as mlflow:
    mlflow.set_experiment("gui_deployable_multiclass")
    experiment = mlflow.get_experiment_by_name("gui_deployable_multiclass")
    with mlflow.start_run(experiment_id=experiment.experiment_id) as run:
        mlflow.sklearn.autolog()
        iris = dataiku.Dataset("iris_prepared").get_dataframe()
        iris = iris.dropna()
        X = iris.drop(["class"], axis=1)
        Y = iris[["class"]]
        clf = svm.SVC(kernel="linear", C=1)
        clf.fit(X, Y)
        mlflow_extension.set_run_inference_info(run._info.run_id, "MULTICLASS", clf.classes_.tolist(),
                                                "WinMLflow_37", "class")

    run = mlflow.get_run(run._info.run_id)  # refresh run
```

- Go in Experiment Tracking. Check that you have a "gui_deployable_multiclass" experiment, with a successful run.
- Deploy the model of the run as a new version of a new model. Use "iris_prepared" as the evaluation dataset
- Create a scoring recipe to score the newly created model on iris_dataset
- Check that the scored dataset looks ok
- Create an evaluate recipe to evaluate the newly created model on iris_dataset, having a Model Evaluation Store as an output
- Check that the single Model Evaluation looks ok (has metrics, including a Data Drift metric)

**Comment by lpenet** [comment]
To test upload is ok on windows, create the following tree:
```
c:\temps\upload_test
  fichier1.txt
  fichier 2.txt
  sub1
    fichier 3.txt
  sub 2
    fichier4.txt
```
Then uploaded it in the experiments managed folder running the following python code:
```python
mf.upload_folder("test_upload", "C:\\Temp\\upload_test")
```

and checked that I can find the content of `upload_test` in the `test_upload` dir of the managed folder.

**Comment by lpenet** [comment]
To test on GNU/Linux:
- manually ran autolog_tests
- DOD of companion test PR is all green https://github.com/dataiku/dip/pull/17453#issuecomment-1209427534
- create a test directory:
```
~tmp/upload_test
  fichier1
  fichier 2
  fichier\3
  sub1
    fichier_4.txt
  sub 2
    fichier5.txt
  sub\3
    fichier6.txt
```

and tested upload in MF is ok.

**Review by lpenet — APPROVED** [review]
LGTM+Tested
